common/src/main/java/io/pravega/common/util/CompositeByteArraySegment.java

```diff
@@ -112,14 +113,126 @@ private CompositeByteArraySegment(Object[] arrays, int arraySize, int startOffse
 
     @Override
     public byte get(int offset) {
-        byte[] array = getArray(getArrayId(offset), false); // No need to allocate array if not allocated yet.
-        return array == null ? 0 : array[getArrayOffset(offset)];
+        ByteBuffer bb = getBuffer(getBufferId(offset), false); // No need to allocate array if not allocated yet.
+        return bb == null ? 0 : bb.array()[getBufferOffset(offset)];
     }
 
     @Override
     public void set(int offset, byte value) {
-        byte[] array = getArray(getArrayId(offset), true); // Need to allocate array if not allocated yet.
-        array[getArrayOffset(offset)] = value;
+        ByteBuffer bb = getBuffer(getBufferId(offset), true); // Need to allocate array if not allocated yet.
+        bb.array()[getBufferOffset(offset)] = value;
+    }
+
+    @Override
+    public void setShort(int offset, short value) {
+        Exceptions.checkArrayRange(offset, Short.BYTES, this.length, "index", "length");
+
+        int bufferId = getBufferId(offset);
+        ByteBuffer bb = getBuffer(bufferId, true);
+        int bufferOffset = getBufferOffset(offset);
+        try {
+            bb.putShort(bufferOffset, value);
+        } catch (IndexOutOfBoundsException | BufferOverflowException ex) {
```

Comment:
Benchmarking has shown that conditional performs 10-15% worse than this. If did a conditional, I would have had to calculate the remaining length every time I invoke this method, which translates to extra CPU cycles.

I am aware this is not super clean code, but it is optimized for the case where the value does fit - as that is by far the most common case. 

IMO, it is ok to be slightly sloppy here. This is extremely low-level code that is invoked very frequenty; any optimization here can speed up upstream code. And viceversa: any slowdown here has bad ramifications upstream.

Reply:
It may be faster in the happy case, but this case it far from impossible and will be tens of thousands of times more costly in the unhappy case. It needs to be a very low probably for a 15% benefit to pay off on average. I see no evidence that this is the case here.

Reply:
I have reworked this class in the following way:
- I'm requiring that all component buffer lengths be a power of 2. This enables me to use bitwise operations to get buffer ids and buffer offsets (instead of expensive divisions and modulos). 
- I created a `BufferLayout`, which in my case, is instantiated once and reused with every `CompositeByteArrayInstance`. This class precalculates the last offsets at which I can write longs, ints and shorts, so all I need to do is compare the offset with those values. That is less expensive than having do do a subtraction for every write.